[GPUP] Remove TrackPrivateRemoteIdentifier,

jyavenard opened the pull request:
#### e0a0bc967262c6e4741e8539ab3c32c3bc310233
<pre>
[GPUP] Remove TrackPrivateRemoteIdentifier,
<a href="https://bugs.webkit.org/show_bug.cgi?id=265718">https://bugs.webkit.org/show_bug.cgi?id=265718</a>
<a href="https://rdar.apple.com/119070875">rdar://119070875</a>

Reviewed by Jer Noble.

With TrackPrivate now using an uint64_t as track identifier,
we no longer require the TrackPrivateRemoteIdentifier object
to bridge the old AtomString trackID between the content and GPU process.
Instead we can use the TrackID int everywhere. The TrackID comes from the
media container that originated from the content process but was demuxed in the
GPU process.

We now fully comply with the HTML spec as giving access to the Audio/Video/Text
track to the container&apos;s track ID is a compulsory requirement:
&quot;The user agent MUST source attribute values for id, kind, label and language for AudioTrack, VideoTrack and TextTrack objects as described for MPEG-4 ISOBMFF in the in-band tracks spec&quot; [1][2]
That the ID of an Audio and Video track is an int is also a requirement:
ISO-BMFF &quot;the decimal representation of the track_ID of a TrackHeaderBox (tkhd) in a TrackBox (trak).&quot; [3]
WebM &quot;Decimal representation of the TrackNumber field of the track in the Track section of the WebM file Segment&quot; [4]
MPEG-TS: &quot;Decimal representation of the elementary stream&apos;s identifier (elementary_PID field) in the PMT.&quot; [5]

In ISO-BMFF, the track_id in a TrackHeaderBox is an unsigned 32 bits int.
In Mpeg-TS, the Elementary PID is a 13 bits unsigned int
In WebM, there is theoretically no limitation to the length of an int as the EBML design allows for variable length int.
However, Matroska does requires integers to be 0-8 bytes in size.
&quot;EBML files use integers of variable size. This way, the file format doesn’t waste space with storing 32 or even 64 bit integers in placed where they might sometimes occur. The way the size is coded is inspired by the UTF-8 encoding format. [...] The Matroska file format does not allow integer lengths greater than 8, meaning that the number of leading zeros is not higher than 7 and that the total length can always be retrieved from the first byte.&quot; [6]

As a consequence, there is no theoretical restriction to the values usable by a TrackID,
making the HashTable unsuitable as storage container as it excludes at least two values
through the design of its hashing algorithm (it requires a value used for an empty element and a deleted one)
So we use WTF::StdUnorderedMap instead.
Using a StdUnorderedMap also provides the advantage that there is no longer a need to validate
the TrackID value prior sending it over IPC as there are no exclusions possible for its use as
key in a HashMap
Should the value not be linked to an existing TrackPrivate it will be ignored.

[1] <a href="https://w3c.github.io/mse-byte-stream-format-isobmff/#iso-init-segments">https://w3c.github.io/mse-byte-stream-format-isobmff/#iso-init-segments</a>
[2] <a href="https://w3c.github.io/mse-byte-stream-format-webm/#webm-init-segmentsA">https://w3c.github.io/mse-byte-stream-format-webm/#webm-init-segmentsA</a>
[3] <a href="https://dev.w3.org/html5/html-sourcing-inband-tracks/#mpeg4">https://dev.w3.org/html5/html-sourcing-inband-tracks/#mpeg4</a>
[4] <a href="https://dev.w3.org/html5/html-sourcing-inband-tracks/#webmA">https://dev.w3.org/html5/html-sourcing-inband-tracks/#webmA</a>
[5] <a href="https://dev.w3.org/html5/html-sourcing-inband-tracks/#mpeg2ts">https://dev.w3.org/html5/html-sourcing-inband-tracks/#mpeg2ts</a>
[6] <a href="https://www.matroska.org/files/matroska_file_format_alexander_noe.pdf">https://www.matroska.org/files/matroska_file_format_alexander_noe.pdf</a>

Fly-By: Fix the content process RemoteTrackPrivate that would never be deleted
when its corresponding TrackPrivateProxy was deleted in the GPU process.
Fly-By: InbandChapterTrackPrivateAVFObjC didn&apos;t have a unique TrackID, we assign
one to it. The code worked with the GPU process as the RemoteTrackPrivate
would have assigned it a unique ID; however following changes in 271419@main,
the original track requires a unique TrackID as the RemoteTrackPrivate reports
the same.

Add test ensuring track ID of value 2^64-1 are playable.

* LayoutTests/media/content/short--1-track.webm: Added.
* LayoutTests/media/media-audio-track-expected.txt: Added.
* LayoutTests/media/media-audio-track.html: Added.
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::processNewAndRemovedTextTracks):
(WebCore::InbandChapterTrackPrivateAVFObjC::create): Add TrackID constructor argument
* Source/WebCore/platform/graphics/avfoundation/objc/InbandChapterTrackPrivateAVFObjC.mm:
(WebCore::InbandChapterTrackPrivateAVFObjC::InbandChapterTrackPrivateAVFObjC):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
Assign a different TrackID whenever a text track is created.
(WebCore::MediaPlayerPrivateAVFoundationObjC::processChapterTracks):
* Source/WebKit/GPUProcess/media/InitializationSegmentInfo.h:
* Source/WebKit/GPUProcess/media/InitializationSegmentInfo.serialization.in:
* Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.cpp:
(WebKit::RemoteAudioTrackProxy::RemoteAudioTrackProxy):
(WebKit::RemoteAudioTrackProxy::configurationChanged):
* Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerManagerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::addRemoteAudioTrackProxy):
(WebKit::RemoteMediaPlayerProxy::audioTrackSetEnabled):
(WebKit::RemoteMediaPlayerProxy::addRemoteVideoTrackProxy):
(WebKit::RemoteMediaPlayerProxy::videoTrackSetSelected):
(WebKit::RemoteMediaPlayerProxy::addRemoteTextTrackProxy):
(WebKit::RemoteMediaPlayerProxy::textTrackSetMode):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerDidRemoveAudioTrack):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerDidRemoveVideoTrack):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerDidRemoveTextTrack):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitializationSegment):
(WebKit::RemoteSourceBufferProxy::addTrackBuffer):
(WebKit::RemoteSourceBufferProxy::updateTrackIds):
(WebKit::RemoteSourceBufferProxy::bufferedSamplesForTrackId):
(WebKit::RemoteSourceBufferProxy::enqueuedSamplesForTrackID):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/GPUProcess/media/RemoteTextTrackProxy.cpp:
(WebKit::RemoteTextTrackProxy::RemoteTextTrackProxy):
(WebKit::RemoteTextTrackProxy::configurationChanged):
(WebKit::RemoteTextTrackProxy::addDataCue):
(WebKit::RemoteTextTrackProxy::updateDataCue):
(WebKit::RemoteTextTrackProxy::removeDataCue):
(WebKit::RemoteTextTrackProxy::addGenericCue):
(WebKit::RemoteTextTrackProxy::updateGenericCue):
(WebKit::RemoteTextTrackProxy::removeGenericCue):
(WebKit::RemoteTextTrackProxy::parseWebVTTFileHeader):
(WebKit::RemoteTextTrackProxy::parseWebVTTCueData):
* Source/WebKit/GPUProcess/media/RemoteTextTrackProxy.h:
* Source/WebKit/GPUProcess/media/RemoteVideoTrackProxy.cpp:
(WebKit::RemoteVideoTrackProxy::RemoteVideoTrackProxy):
(WebKit::RemoteVideoTrackProxy::updateConfiguration):
* Source/WebKit/GPUProcess/media/RemoteVideoTrackProxy.h:
* Source/WebKit/GPUProcess/media/TextTrackPrivateRemoteConfiguration.serialization.in:
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteConfiguration.h:
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteConfiguration.serialization.in:
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteIdentifier.h: Removed.
* Source/WebKit/Scripts/webkit/messages.py:
(function_parameter_type):
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.cpp:
(WebKit::AudioTrackPrivateRemote::AudioTrackPrivateRemote):
(WebKit::AudioTrackPrivateRemote::setEnabled):
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::addRemoteAudioTrack):
(WebKit::MediaPlayerPrivateRemote::removeRemoteAudioTrack):
(WebKit::MediaPlayerPrivateRemote::remoteAudioTrackConfigurationChanged):
(WebKit::MediaPlayerPrivateRemote::addRemoteTextTrack):
(WebKit::MediaPlayerPrivateRemote::removeRemoteTextTrack):
(WebKit::MediaPlayerPrivateRemote::remoteTextTrackConfigurationChanged):
(WebKit::MediaPlayerPrivateRemote::parseWebVTTFileHeader):
(WebKit::MediaPlayerPrivateRemote::parseWebVTTCueData):
(WebKit::MediaPlayerPrivateRemote::parseWebVTTCueDataStruct):
(WebKit::MediaPlayerPrivateRemote::addDataCue):
(WebKit::MediaPlayerPrivateRemote::addDataCueWithType):
(WebKit::MediaPlayerPrivateRemote::updateDataCue):
(WebKit::MediaPlayerPrivateRemote::removeDataCue):
(WebKit::MediaPlayerPrivateRemote::addGenericCue):
(WebKit::MediaPlayerPrivateRemote::updateGenericCue):
(WebKit::MediaPlayerPrivateRemote::removeGenericCue):
(WebKit::MediaPlayerPrivateRemote::addRemoteVideoTrack):
(WebKit::MediaPlayerPrivateRemote::removeRemoteVideoTrack):
(WebKit::MediaPlayerPrivateRemote::remoteVideoTrackConfigurationChanged):
(WebKit::MediaPlayerPrivateRemote::audioTrackPrivateRemote const):
(WebKit::MediaPlayerPrivateRemote::videoTrackPrivateRemote const):
(WebKit::MediaPlayerPrivateRemote::textTrackPrivateRemote const):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.messages.in:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::addTrackBuffer):
(WebKit::SourceBufferPrivateRemote::updateTrackIds):
(WebKit::SourceBufferPrivateRemote::bufferedSamplesForTrackId):
(WebKit::SourceBufferPrivateRemote::enqueuedSamplesForTrackID):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateDidReceiveInitializationSegment):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.cpp:
(WebKit::TextTrackPrivateRemote::TextTrackPrivateRemote):
(WebKit::TextTrackPrivateRemote::setMode):
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.cpp:
(WebKit::VideoTrackPrivateRemote::VideoTrackPrivateRemote):
(WebKit::VideoTrackPrivateRemote::setSelected):
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.h:
(WebKit::VideoTrackPrivateRemote::create):

Canonical link: <a href="https://commits.webkit.org/271599@main">https://commits.webkit.org/271599@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/294140223984de8a111a06e0467a4a4cc3cb0890

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28930 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7576 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30286 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31545 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26367 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29503 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4918 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26398 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29202 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6265 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24836 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5461 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/29091 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5581 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25861 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32882 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26457 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26293 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5562 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3735 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29612 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7204 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/25655 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6915 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6047 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6083 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebKit/GPUProcess/media/InitializationSegmentInfo.h:
```diff
@@ -39,7 +38,7 @@ struct InitializationSegmentInfo {
     struct TrackInformation {
         MediaDescriptionInfo description;
-        TrackPrivateRemoteIdentifier identifier;
+        uint64_t id;
```
Shouldn't this be a `WebCore::TrackID`?

jernoble commented on Source/WebKit/GPUProcess/media/InitializationSegmentInfo.serialization.in:
```diff
@@ -31,7 +31,7 @@ struct WebKit::InitializationSegmentInfo {
 [Nested] struct WebKit::InitializationSegmentInfo::TrackInformation {
     WebKit::MediaDescriptionInfo description;
-    WebKit::TrackPrivateRemoteIdentifier identifier;
+    uint64_t id;
```
Ditto.

jyavenard replied:
this require to either create a TrackID.h or modify the python generator to use the header where it's defined..

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.cpp:
```diff
@@ -75,7 +75,8 @@ void RemoteAudioTrackProxy::configurationChanged()
 
-    m_connectionToWebProcess->connection().send(Messages::MediaPlayerPrivateRemote::RemoteAudioTrackConfigurationChanged(m_identifier, configuration()), m_mediaPlayerIdentifier);
+    m_connectionToWebProcess->connection().send(Messages::MediaPlayerPrivateRemote::RemoteAudioTrackConfigurationChanged(m_id, configuration()), m_mediaPlayerIdentifier);
+    m_id = m_trackPrivate->id();
```
This looks like it's in the wrong order. Shouldn't `m_id` be reset _before_ sending the message?

jyavenard replied:
No. Because we send the old id to the content process along the new one. It can the. Determine if the Id changed

jyavenard replied:
I'll use std::exchange to make the code clearer.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.h:
```diff
@@ -50,22 +49,23 @@ class RemoteAudioTrackProxy final
         m_enabled = enabled;
         m_trackPrivate->setEnabled(enabled);
     }
+    bool operator==(const WebCore::AudioTrackPrivate& track) const { return track == m_trackPrivate.get(); }
```
Do you really want to check the equality of a `RemoteAudioTrackProxy` and a `AudioTrackPrivate`? It's not clear where you're using this from.

jernoble replied:
I see now where this is being used from. It still seems a little weird, but I guess it's ok.

jyavenard replied:
I thought that it made the code more elegant where it's used

jernoble commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -590,100 +590,97 @@ void RemoteMediaPlayerProxy::requestHostingContextID(LayerHostingContextIDCallba
+            return;
+        if (m_videoTracks[index]->id() == track.id()) {
+            m_videoTracks.remove(index);
+            break;
```
Why not just create the replacement track and assign it here?

jernoble commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -590,100 +590,97 @@ void RemoteMediaPlayerProxy::requestHostingContextID(LayerHostingContextIDCallba
+            return;
+        if (m_textTracks[index]->id() == track.id()) {
+            m_textTracks.remove(index);
+            break;
```
Ditto.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -697,13 +694,11 @@ void RemoteMediaPlayerProxy::mediaPlayerDidAddAudioTrack(WebCore::AudioTrackPriv
+    for (size_t index = 0; index < m_audioTracks.size(); index++) {
+        if (m_audioTracks[index]->id() == track.id())
+            m_audioTracks.remove(index);
+    }
```
Vector already has a `removeFirstMatching(...)` function that makes this for loop unnecessary.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -713,13 +708,11 @@ void RemoteMediaPlayerProxy::mediaPlayerDidAddVideoTrack(WebCore::VideoTrackPriv
+    for (size_t index = 0; index < m_videoTracks.size(); index++) {
+        if (m_videoTracks[index]->id() == track.id())
+            m_videoTracks.remove(index);
+    }
```
Ditto.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -729,13 +722,11 @@ void RemoteMediaPlayerProxy::mediaPlayerDidAddTextTrack(WebCore::InbandTextTrack
+    for (size_t index = 0; index < m_textTracks.size(); index++) {
+        if (m_textTracks[index]->id() == track.id())
+            m_textTracks.remove(index);
+    }
```
Ditto.

jernoble commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	archiveVersion = 1;
 	classes = {
 	};
-	objectVersion = 55;
```
Probably don't want to commit this.

cdumez commented on Source/WebKit/GPUProcess/media/InitializationSegmentInfo.serialization.in:
```diff
@@ -31,7 +31,7 @@ struct WebKit::InitializationSegmentInfo {
 [Nested] struct WebKit::InitializationSegmentInfo::TrackInformation {
     WebKit::MediaDescriptionInfo description;
-    WebKit::TrackPrivateRemoteIdentifier identifier;
+    uint64_t id;
```
We shouldn't be passing non-strongly typed identifiers via IPC. This change seems like a regression security wise?
Ids sent over IPC need to be validated because they're often used to look up in HashMaps and such and lead to crashes on recipient side.

jyavenard replied:
There's heaps of places where we are passing int across though when it comes to selected track, track index, number of frames. This made it consistent.

To me if we want to assess that the int is a valid one, it's the wrong place to start.

Here, every path is tested to check if we have a new track and it will safely abort

jyavenard replied:
additionally, what we want to pass is the track ID as is found in the original container, which is an int. Storing that int as an identifier, pass it over IPC only to read it as an int again (after looking it up in a HashMap) seems wasteful.

> they're often used to look up in HashMaps and such and lead to crashes on recipient side.

Looking up the TrackPrivateRemoteIdentifier in a map to find the original ID is exactly what it was doing before, so we removed possibility of mis-used, not add some.

cdumez replied:
The TrackPrivateRemoteIdentifier that was received over IPC was already validated at the IPC layer. If it were an identifier that could not be looked up in a HashMap, IPC decoding would have failed. So what you call wasteful is actually valuable security, which we worked to achieve.

Now that you're passing an int, we no longer get ANY IPC validation. This is not acceptable. I think this patch is a step in the wrong direction.

jyavenard replied:
I provided further details in the commit log as to the decision of no longer using HashMap and how there is nothing to validate as the ID is never invalid. Please reconsider.

cdumez replied:
I removed my objection given that I don't have a better proposal at the moment given the id requirements. @achristensen07 may have an idea though.

jyavenard replied:
I added a test with a file that is using 2^64-1 as track id

jyavenard committed:
[GPUP] Remove TrackPrivateRemoteIdentifier,
https://bugs.webkit.org/show_bug.cgi?id=265718
rdar://119070875

Reviewed by Jer Noble.

With TrackPrivate now using an uint64_t as track identifier,
we no longer require the TrackPrivateRemoteIdentifier object
to bridge the old AtomString trackID between the content and GPU process.
Instead we can use the TrackID int everywhere. The TrackID comes from the
media container that originated from the content process but was demuxed in the
GPU process.

We now fully comply with the HTML spec as giving access to the Audio/Video/Text
track to the container's track ID is a compulsory requirement:
"The user agent MUST source attribute values for id, kind, label and language for AudioTrack, VideoTrack and TextTrack objects as described for MPEG-4 ISOBMFF in the in-band tracks spec" [1][2]
That the ID of an Audio and Video track is an int is also a requirement:
ISO-BMFF "the decimal representation of the track_ID of a TrackHeaderBox (tkhd) in a TrackBox (trak)." [3]
WebM "Decimal representation of the TrackNumber field of the track in the Track section of the WebM file Segment" [4]
MPEG-TS: "Decimal representation of the elementary stream's identifier (elementary_PID field) in the PMT." [5]

In ISO-BMFF, the track_id in a TrackHeaderBox is an unsigned 32 bits int.
In Mpeg-TS, the Elementary PID is a 13 bits unsigned int
In WebM, there is theoretically no limitation to the length of an int as the EBML design allows for variable length int.
However, Matroska does requires integers to be 0-8 bytes in size.
"EBML files use integers of variable size. This way, the file format doesn’t waste space with storing 32 or even 64 bit integers in placed where they might sometimes occur. The way the size is coded is inspired by the UTF-8 encoding format. [...] The Matroska file format does not allow integer lengths greater than 8, meaning that the number of leading zeros is not higher than 7 and that the total length can always be retrieved from the first byte." [6]

As a consequence, there is no theoretical restriction to the values usable by a TrackID,
making the HashTable unsuitable as storage container as it excludes at least two values
through the design of its hashing algorithm (it requires a value used for an empty element and a deleted one)
So we use WTF::StdUnorderedMap instead.
Using a StdUnorderedMap also provides the advantage that there is no longer a need to validate
the TrackID value prior sending it over IPC as there are no exclusions possible for its use as
key in a HashMap
Should the value not be linked to an existing TrackPrivate it will be ignored.

[1] https://w3c.github.io/mse-byte-stream-format-isobmff/#iso-init-segments
[2] https://w3c.github.io/mse-byte-stream-format-webm/#webm-init-segmentsA
[3] https://dev.w3.org/html5/html-sourcing-inband-tracks/#mpeg4
[4] https://dev.w3.org/html5/html-sourcing-inband-tracks/#webmA
[5] https://dev.w3.org/html5/html-sourcing-inband-tracks/#mpeg2ts
[6] https://www.matroska.org/files/matroska_file_format_alexander_noe.pdf

Fly-By: Fix the content process RemoteTrackPrivate that would never be deleted
when its corresponding TrackPrivateProxy was deleted in the GPU process.
Fly-By: InbandChapterTrackPrivateAVFObjC didn't have a unique TrackID, we assign
one to it. The code worked with the GPU process as the RemoteTrackPrivate
would have assigned it a unique ID; however following changes in 271419@main,
the original track requires a unique TrackID as the RemoteTrackPrivate reports
the same.

Add test ensuring track ID of value 2^64-1 are playable.

* LayoutTests/media/content/short--1-track.webm: Added.
* LayoutTests/media/media-audio-track-expected.txt: Added.
* LayoutTests/media/media-audio-track.html: Added.
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::processNewAndRemovedTextTracks):
(WebCore::InbandChapterTrackPrivateAVFObjC::create): Add TrackID constructor argument
* Source/WebCore/platform/graphics/avfoundation/objc/InbandChapterTrackPrivateAVFObjC.mm:
(WebCore::InbandChapterTrackPrivateAVFObjC::InbandChapterTrackPrivateAVFObjC):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
Assign a different TrackID whenever a text track is created.
(WebCore::MediaPlayerPrivateAVFoundationObjC::processChapterTracks):
* Source/WebKit/GPUProcess/media/InitializationSegmentInfo.h:
* Source/WebKit/GPUProcess/media/InitializationSegmentInfo.serialization.in:
* Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.cpp:
(WebKit::RemoteAudioTrackProxy::RemoteAudioTrackProxy):
(WebKit::RemoteAudioTrackProxy::configurationChanged):
* Source/WebKit/GPUProcess/media/RemoteAudioTrackProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerManagerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::addRemoteAudioTrackProxy):
(WebKit::RemoteMediaPlayerProxy::audioTrackSetEnabled):
(WebKit::RemoteMediaPlayerProxy::addRemoteVideoTrackProxy):
(WebKit::RemoteMediaPlayerProxy::videoTrackSetSelected):
(WebKit::RemoteMediaPlayerProxy::addRemoteTextTrackProxy):
(WebKit::RemoteMediaPlayerProxy::textTrackSetMode):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerDidRemoveAudioTrack):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerDidRemoveVideoTrack):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerDidRemoveTextTrack):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitializationSegment):
(WebKit::RemoteSourceBufferProxy::addTrackBuffer):
(WebKit::RemoteSourceBufferProxy::updateTrackIds):
(WebKit::RemoteSourceBufferProxy::bufferedSamplesForTrackId):
(WebKit::RemoteSourceBufferProxy::enqueuedSamplesForTrackID):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/GPUProcess/media/RemoteTextTrackProxy.cpp:
(WebKit::RemoteTextTrackProxy::RemoteTextTrackProxy):
(WebKit::RemoteTextTrackProxy::configurationChanged):
(WebKit::RemoteTextTrackProxy::addDataCue):
(WebKit::RemoteTextTrackProxy::updateDataCue):
(WebKit::RemoteTextTrackProxy::removeDataCue):
(WebKit::RemoteTextTrackProxy::addGenericCue):
(WebKit::RemoteTextTrackProxy::updateGenericCue):
(WebKit::RemoteTextTrackProxy::removeGenericCue):
(WebKit::RemoteTextTrackProxy::parseWebVTTFileHeader):
(WebKit::RemoteTextTrackProxy::parseWebVTTCueData):
* Source/WebKit/GPUProcess/media/RemoteTextTrackProxy.h:
* Source/WebKit/GPUProcess/media/RemoteVideoTrackProxy.cpp:
(WebKit::RemoteVideoTrackProxy::RemoteVideoTrackProxy):
(WebKit::RemoteVideoTrackProxy::updateConfiguration):
* Source/WebKit/GPUProcess/media/RemoteVideoTrackProxy.h:
* Source/WebKit/GPUProcess/media/TextTrackPrivateRemoteConfiguration.serialization.in:
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteConfiguration.h:
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteConfiguration.serialization.in:
* Source/WebKit/GPUProcess/media/TrackPrivateRemoteIdentifier.h: Removed.
* Source/WebKit/Scripts/webkit/messages.py:
(function_parameter_type):
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.cpp:
(WebKit::AudioTrackPrivateRemote::AudioTrackPrivateRemote):
(WebKit::AudioTrackPrivateRemote::setEnabled):
* Source/WebKit/WebProcess/GPU/media/AudioTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::addRemoteAudioTrack):
(WebKit::MediaPlayerPrivateRemote::removeRemoteAudioTrack):
(WebKit::MediaPlayerPrivateRemote::remoteAudioTrackConfigurationChanged):
(WebKit::MediaPlayerPrivateRemote::addRemoteTextTrack):
(WebKit::MediaPlayerPrivateRemote::removeRemoteTextTrack):
(WebKit::MediaPlayerPrivateRemote::remoteTextTrackConfigurationChanged):
(WebKit::MediaPlayerPrivateRemote::parseWebVTTFileHeader):
(WebKit::MediaPlayerPrivateRemote::parseWebVTTCueData):
(WebKit::MediaPlayerPrivateRemote::parseWebVTTCueDataStruct):
(WebKit::MediaPlayerPrivateRemote::addDataCue):
(WebKit::MediaPlayerPrivateRemote::addDataCueWithType):
(WebKit::MediaPlayerPrivateRemote::updateDataCue):
(WebKit::MediaPlayerPrivateRemote::removeDataCue):
(WebKit::MediaPlayerPrivateRemote::addGenericCue):
(WebKit::MediaPlayerPrivateRemote::updateGenericCue):
(WebKit::MediaPlayerPrivateRemote::removeGenericCue):
(WebKit::MediaPlayerPrivateRemote::addRemoteVideoTrack):
(WebKit::MediaPlayerPrivateRemote::removeRemoteVideoTrack):
(WebKit::MediaPlayerPrivateRemote::remoteVideoTrackConfigurationChanged):
(WebKit::MediaPlayerPrivateRemote::audioTrackPrivateRemote const):
(WebKit::MediaPlayerPrivateRemote::videoTrackPrivateRemote const):
(WebKit::MediaPlayerPrivateRemote::textTrackPrivateRemote const):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.messages.in:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::addTrackBuffer):
(WebKit::SourceBufferPrivateRemote::updateTrackIds):
(WebKit::SourceBufferPrivateRemote::bufferedSamplesForTrackId):
(WebKit::SourceBufferPrivateRemote::enqueuedSamplesForTrackID):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateDidReceiveInitializationSegment):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.cpp:
(WebKit::TextTrackPrivateRemote::TextTrackPrivateRemote):
(WebKit::TextTrackPrivateRemote::setMode):
* Source/WebKit/WebProcess/GPU/media/TextTrackPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.cpp:
(WebKit::VideoTrackPrivateRemote::VideoTrackPrivateRemote):
(WebKit::VideoTrackPrivateRemote::setSelected):
* Source/WebKit/WebProcess/GPU/media/VideoTrackPrivateRemote.h:
(WebKit::VideoTrackPrivateRemote::create):

Canonical link: https://commits.webkit.org/271599@main
